### ISSUE TITLE
smart crop to optimize training on regions on interest

### ISSUE THREAD

**Issue by rak5216** [mit-quest/necstlab-damage-segmentation#43]
cropping algorithm that prioritizes regions of high class representation as well as sparse regions

**Comment by rak5216** [comment]
Proposed start (@Josh-Joseph): 
Case 1 (alternate between crops with and without given class): Class 1: Find all crops, split them into those with Class 1 present and those without Class 1 present. First, take crop randomly sampled from crop group WITH Class 1 present (record crop). Second (if desired), take crop randomly sampled from group WITHOUT Class 1 present (record). Third (if desired), take crop randomly sampled from crop group WITH Class 1 present (record). Fourth (if desired), take crop randomly sampled from group WITHOUT Class 1 present (record). ETC. Move to Class 2. Repeat process used for Class 1, but ensure no duplicates. Repeat random selection for Class 2 until no duplicates vs. any other class cropping for given image. If all crops with given class present are taken, even though you are looking for a crop WITH a given class, take a crop from the group without the class to avoid duplicates. Vice versa when looking for WITHOUT class but all these crops are already taken. Break loop once all possible crops are taken (for ESRF images, there are 20 crops possible), regardless of `num_per_class` value in config file.

**Comment by Josh-Joseph** [comment]
The main change I'd suggest is to not worry about what crops another class used and accept there will be some overlap and that is fine. So I'd create each class's crops the same way. Here's the pseudo-ish code I'd suggest:
- for each class
  - sample `num_positive` images with the class present
  - sample `num_negative` images without the class present

I think some intra-class duplication is fine here but we are probably going to want the standard random sampling for validation and test so it's more "fair" and not skewed by this sampling. Thoughts?

**Comment by Josh-Joseph** [comment]
An alternate, and much fancier version, is to sample *all* non-overlapping images with damage -- how ever many that is.

**Comment by rak5216** [comment]
> An alternate, and much fancier version, is to sample _all_ non-overlapping images with damage -- how ever many that is.

With the exception on the "non-overlapping" criterion, I think this refers to the `all` crop option already enabled. For the current images, that's 20 crops per image. For the larger micro-CT images, that's around 35 crops. The current `all` method allows overlaps at the RHS and bottom edges to avoid training on padded crops. Could easily adapt this for a non-overlap option though, but hard to say if the region skipped at the RHS and bottom edges is valuable, this is scan dependent.

**Comment by rak5216** [comment]
> The main change I'd suggest is to not worry about what crops another class used and accept there will be some overlap and that is fine. So I'd create each class's crops the same way. Here's the pseudo-ish code I'd suggest:
> 
> * for each class
>   
>   * sample `num_positive` images with the class present
>   * sample `num_negative` images without the class present
> 
> I think some intra-class duplication is fine here but we are probably going to want the standard random sampling for validation and test so it's more "fair" and not skewed by this sampling. Thoughts?

Is the concern for avoiding crop duplication from class to class that it will be difficult to implement? I think I can do it, though I'm still playing around with assembling 3D arrays comprised of 1D crop # (counter ID) and 2D crop, and then pulling a 2D crop from 3D array based the crop ID. Once I get this it will just be a series of conditional statements checking for duplication and maybe a `while` loop. If this works and there's no duplication in crop selection, which might be characterized as random but weighted by class presence), can we deploy this same selection procedure for val and test sets too? Seems like duplication is what deems "unfair", though maybe prioritizing training for class presence also is unfair outside of training.

**Comment by Josh-Joseph** [comment]
I think my main concern is the difficulty in may introduce in interpreting results. So let's imagine we have 45 degree ply damage and fiber break damage. And if (..I'm making stuff up here) there is a *lot* of fiber break damage it is included as the first class this may block all the 45 degree ply damage crops.

This sort of dependency where the performance of the classes may depend on the order we list them can sometimes create real confusing things. That said, if you think this won't be a big issue and is important happy to defer to you there

**Comment by rak5216** [comment]
My thought is that if a fiber break-heavy crop also happens to be a 45-dmg-heavy crop, then it shouldn't matter whether this crop is selected when prioritizing the former or the latter class. The important thing is that it's even selected. When 45 get its turn (assuming listed second), then it may want to pick the same crop randomly, BUT a conditional statement will say pick another since it's already selected. So I think this only benefits the training process, since now two good 45 crops are chosen instead of the same one twice. If all good 45 crops are selected during the dominant fiber break selection process, then this is still great because all 45 crops were indeed chosen (and these crops will train for each class simultaneously). Then when it's time for 45-based selection, all selections for class-positive crops will be converted to selections for class-negative crops to avoid duplication and further diversify the training set (seems like no harm done here). Once all crops are selected (total # based on my `all` crops code), independent of which class selects them, then the selection process for the current image stops, and the next image is considered. Overall, for a given crop, since all classes are trained on simultaneously, then the order of class listing and how said order affects the crop selection I don't think matters for class performance. But my whole thought process is based on the current fact that crop selection identically chooses the crop for the raw image and the annotated image. Then the annotated image is decomposed into class masks that share the crop position.

**Comment by rak5216** [comment]
Though, a big thing with my plan is that each image may a slightly different number of crops depending on class presence, due to no duplication. Which I don't think presents a fairness or generalization problem, but I'd have to think about that. Seems only way to guarantee same # crops per img is to allow duplication

**Comment by rak5216** [comment]
I guess duplication may be okay due to the random rotations augmenting duplicates.  hmmm

**Comment by Josh-Joseph** [comment]
Before we answer this maybe what's important is to rehash a bit the purpose here -- in my view the purpose of addressing this issue is to speed up training by including more damage/image (and therefore needing less images which will speed up training). Was that your picture of the main purpose also or did you view this differently?

**Comment by rak5216** [comment]
Yes totally agree on purpose, I think I way overthought this. 

So now I've come to a different outlook that may also improve fairness. A downside, which I didn't see before, possibly for the current `linear` or `all` cropping methods is how the crops are "rigidly" positioned in the image. For these cases, I simply begin at top left and sample fully across the img without overlap until I reach the RHS. Then I move down by target size and repeat until reaching RHS. The last crop for `all` is always bounded by the image RHS and bottom image edges (overlap allowed). The problem here is that for the combination of some cracks, like 0 cracks, and scan orientations (like those where the 90 ply is blurry, a 0 crack may seem to appear in the same position in each crop, since the position doesn't move as one scrolls through slices. In contrast, if you scroll thru slices, 45 cracks will appear to move left to right. So I wouldn't want to accidentally train a model to think a 0 crack will always appear some distance from the crop edge. This crop sampling was the basis of my former arguments, which I now think undermines it completely.

Thus, I think the easiest way to accomplish the purpose to randomly sample a crop as before, but now check to see if given class is present (or not). If the class is present in the random crop, good, then move on. If the class is not present, but i want it to be, pick another random sample and repeat this until I find a crop with it present. Vice versa for searching for crops without a class. Now, I won't have to figure out how to track already selected crops and avoid duplication, as duplication is allowed here, but has minimal probability due to random sampling of crop position instead of sampling "rigid" locations.

**Comment by Josh-Joseph** [comment]
Gotcha. Makes sense -- only bit I'm curious about here is how are you thinking about stopping if none of that class is present so it just doesn't keep sampling forever?

**Comment by rak5216** [comment]
Ah great find, that would've brought certain doom even in the small sample set. For each class, I can first check whether it's present in the full-size annotation before proceeding to find it via random sampling.  The latter case is really challenging, need to think

**Comment by rak5216** [comment]
@Josh-Joseph , okay this appears to work as intended now. One other consideration is sometimes literally one pixel will drive a class-positive selection. So depending on application, maybe we want to define class-positive by a minimum threshold of class-positive pixels per crop

**Comment by rak5216** [comment]
“small_3cropClass_1class” description: testing 3 crops (2 pos, 1 neg) per class for 1 class
![MontageImages](https://user-images.githubusercontent.com/55461758/71219175-289a6c80-2292-11ea-80ed-1814d348c0c6.png)
![MontageClass0](https://user-images.githubusercontent.com/55461758/71219179-2b955d00-2292-11ea-8a5a-a3d491257de1.png)

**Comment by rak5216** [comment]
Case 2: “small_3cropClass_3class_min10px” description: testing 3 crops (2 pos, 1 neg) per class for 3 classes. Defined as class-positive crop if min of 10 class-positive pixels found in given crop. (reduce noise).

-->Interestingly, train/val/test # crops was 21/23/21 (each with max of 30). Val shown below.

-->1st montage is raw inputs, 2nd montage is corresponding class_0 masks, 3rd montage is corresponding class_1 masks, 4th montage is corresponding class_2 masks (open images to zoom-in)

![MontageImage](https://user-images.githubusercontent.com/55461758/71220680-6e0d6880-2297-11ea-9ca1-2296458aa4dd.png)
![Montage_class_0](https://user-images.githubusercontent.com/55461758/71220682-706fc280-2297-11ea-8da6-012c6e977358.png)
![Montage_class_1](https://user-images.githubusercontent.com/55461758/71220686-72398600-2297-11ea-9cdc-c8356135d68e.png)
![Montage_class_2](https://user-images.githubusercontent.com/55461758/71220692-75cd0d00-2297-11ea-9d89-2cd3c25686cd.png)

**Comment by Josh-Joseph** [comment]
cool this looks great!

**Comment by rak5216** [comment]
Test Case 3: “small_2cropClass_3class_VarMinPx” description: testing 2 crops (1 pos, 1 neg) per class for 3 classes. Defined as class-positive crop if min_num_class_pos_px: [100, 200, 50] for class_0, _1, _2.

Shown below is test set, only 2 images represented here.

-->1st montage is raw inputs, 2nd montage is corresponding class_0 masks, 3rd montage is corresponding class_1 masks, 4th montage is corresponding class_2 masks (open images to zoom-in)

![MontageImg](https://user-images.githubusercontent.com/55461758/71284075-e58ed680-232f-11ea-964a-dbed9bcf7a60.png)

![Montage_class_0](https://user-images.githubusercontent.com/55461758/71284084-e9225d80-232f-11ea-88b6-c778233e08c3.png)

![Montage_class_1](https://user-images.githubusercontent.com/55461758/71284088-ec1d4e00-232f-11ea-9ed0-c00e5e0cdb02.png)

![Montage_class_2](https://user-images.githubusercontent.com/55461758/71284092-ef183e80-232f-11ea-9887-1b83428fd335.png)

**Comment by rak5216** [comment]
@Josh-Joseph , so it turns out that the code was fine for smart crop with variable minimums. The issue was in loading the yaml. When I load in the list representing min px thresholds, which is written as [100, 200, 50], the yaml loader instead gives [200, 100, 50], so the ordering isn't preserved. Figuring out now how to preserve list order when loading yaml

**Comment by rak5216** [comment]
Thinking of other areas where this could present an issue. If target size was [x, y], the yaml loader might bring in target size [y, x]. Obviously fine if x==y, but we should guarantee preserved ordering

**Comment by Josh-Joseph** [comment]
That's real odd since the yaml array *should* preserve order (eg https://stackoverflow.com/questions/33126045/pyyaml-is-order-of-a-list-preserved). Maybe it's an issue with the loader?

**Comment by rak5216** [comment]
> That's real odd since the yaml array _should_ preserve order (eg https://stackoverflow.com/questions/33126045/pyyaml-is-order-of-a-list-preserved). Maybe it's an issue with the loader?

"What is not preserved in YAML normally is the order of keys in mappings". Most everything else I read supported this too.

**Comment by rak5216** [comment]
I found a safer to do it which I'm testing on the VM now, which leverages the current `class_name` to ensure things are in sync. It worked in my micro-tests on my machine at least. The config file is more intuitive too

**Comment by rak5216** [comment]
Oddly, after including some txt file output to check the yaml loading values, the order is preserved up until this point. So perhaps order is okay after all, and something else was wrong before I made the recent changes.

![image](https://user-images.githubusercontent.com/55461758/71310003-b8e4c880-23dc-11ea-80c8-038dc91dfa7f.png)

**Comment by rak5216** [comment]
@Josh-Joseph  UPDATE: Finally understand why listing `min_num_class_pos` values in separate keys worked (see `I found a safer way to do it which I'm testing on the VM now, which leverages the current class_name to ensure things are in sync. It worked in my micro-tests on my machine at least. The config file is more intuitive too`), and at the same time, why originally representing all `min_num_class_pos` values in a single list did NOT work.

I thought the issue was the list losing its order during yaml loading. The order is preserved I found. However, as noted above from various forums, the order of `class_<#>_annotation_GVs` (since they are keys) is NOT preserved. Note, the ordering is set at `for c, gvs_in_c in class_annotation_mapping.items():`. Sometimes the for loop performs class_2, _1, _0, sometimes _0, _2, _1, etc. Indeed, the ordering is random, as it changes each time I run the code. So since class ordering is random in the for loop for a given program run, this obviously doesn't work with a fixed-order vector representing `min_num_class_pos`, unless the random class order happens to be _0, _1, _2. So the revised `prepare_dataset.py` and `configs/dataset-small_2cropClass_3class_VarMinPx_final.yaml` codes that I will now merge calls everything by `class_name`, so that `min_num_class_pos` always matches the current class.

I ran some debugging with the revised code (a little confusing out of context), so that you can see the random class ordering (here, class_1, _0, _2 is the order; a class is printed only if the current annotation contains it):
class_1_min-200_GV-[175]
Press Enter to continue...
class_0_min-100_GV-[100]
Press Enter to continue...
class_1_min-200_GV-[175]
Press Enter to continue...
class_0_min-100_GV-[100]
Press Enter to continue...
class_2_min-50_GV-[250]
Press Enter to continue...
class_1_min-200_GV-[175]
Press Enter to continue...
class_0_min-100_GV-[100]
Press Enter to continue...
class_0_min-100_GV-[100]
Press Enter to continue...
class_2_min-50_GV-[250]
Press Enter to continue...
Etc.

**Comment by rak5216** [comment]
Closing now, but feel free to reopen if more discussion wanted. Also merging #45 branch
